### PR TITLE
fix: bug fixes on enforce function, pull requests, required reviewers, etc

### PR DESCRIPTION
**Pull request by josh-garcia-gl**
- improve logging
- fix null error on required reviewers conditional
- fix pull request race condition
- add exclude logic to enforce function
- refresh app installation token before expiration

### CODE REVIEW

File: core/src/main/kotlin/me/frmr/github/repopolicy/core/PolicyEngine.kt [hunk 1]
```diff
@@ -96,7 +100,7 @@ class PolicyEngine(val policy: PolicyDescription, val logging: Boolean) {
       q += "org:${subject.org} "
     }
 
-    return searchRequest.q(q).list()
+    return searchRequest.q(q).list().toList()
```
Comment (frmrm) [inline core/src/main/kotlin/me/frmr/github/repopolicy/core/PolicyEngine.kt:103]: Any concerns about how this is going to scale? We're effectively forfeiting pagination and loading all the results in memory at once.

File: core/src/main/kotlin/me/frmr/github/repopolicy/core/PolicyEngine.kt [hunk 2]
```diff
@@ -132,12 +141,23 @@ class PolicyEngine(val policy: PolicyDescription, val logging: Boolean) {
     return policy.rules.flatMap { rule ->
       val repos = findMatchingRepos(rule.subject) ?: emptyList()
       repos.flatMap { repo ->
+        // refresh app installation token if necessary
+        if (this.appInstallationTokenTtl > 0 && this.appInstallationTokenTtl < System.currentTimeMillis()) {
+          initGithubClient()
```
Comment (frmrm) [inline core/src/main/kotlin/me/frmr/github/repopolicy/core/PolicyEngine.kt:146]: There's got to be a better way to handle this than just copy pastaing this block around but it's not occurring to me at the moment

**Review by frmrm — COMMENTED** [review]
Should we have some changes to tests as we go about finding and fixing these bugs? I'm not normally dogmatic on TDD but I'm wondering if we should be more so here.

File: core/src/main/kotlin/me/frmr/github/repopolicy/core/PolicyEngine.kt (hunk 1 above)
Comment (josh-garcia-gl) [reply core/src/main/kotlin/me/frmr/github/repopolicy/core/PolicyEngine.kt:103]: Yeah, I thought about that. On the one hand, if we use pagination we can't have a long running (over 1 hour) job without adding a little more complexity to that method. On the other hand, there could be scale problems if we are dealing with 1000s of repos.

File: core/src/main/kotlin/me/frmr/github/repopolicy/core/PolicyEngine.kt (hunk 2 above)
Comment (josh-garcia-gl) [reply core/src/main/kotlin/me/frmr/github/repopolicy/core/PolicyEngine.kt:146]: I'll take a look at cleaning this up a little bit.

**Comment by josh-garcia-gl** [comment]
> Should we have some changes to tests as we go about finding and fixing these bugs? I'm not normally dogmatic on TDD but I'm wondering if we should be more so here.

1. WRT "fix null error on required reviewers conditional", the weird thing with that is that it should not be `null` due to the type — but sometimes it just doesn’t exist in which case it is `null`. It’s a tricky thing to unit test. I did clean that logic up a little bit to be more explicit though. 
2. I added tests for the "fix pull request race condition", basically we need to have at least one merge setting allowed.

**Review by farmdawgnation — APPROVED** [review]

**Pull request merged**